src/zenml/utils/pipeline_docker_image_builder.py

```diff
@@ -465,8 +483,99 @@ def _gather_requirements_files(
                     ", ".join(f"`{r}`" for r in integration_requirements_list),
                 )
 
+        # Generate requirements files for all ZenML Hub plugins
+        if docker_settings.required_hub_plugins:
+            (
+                hub_internal_requirements,
+                hub_pypi_requirements,
+            ) = PipelineDockerImageBuilder._get_hub_requirements(
+                docker_settings.required_hub_plugins
+            )
+
+            # Plugin packages themselves
+            for i, (index, packages) in enumerate(
+                hub_internal_requirements.items()
+            ):
+                file_name = f".zenml_hub_internal_requirements_{i}"
+                file_lines = [f"-i {index}", *packages]
+                file_contents = "\n".join(file_lines)
+                requirements_files.append((file_name, file_contents))
+                if log:
+                    logger.info(
+                        "- Including internal hub packages from index `%s`: %s",
+                        index,
+                        ", ".join(f"`{r}`" for r in packages),
+                    )
+
+            # PyPI requirements of plugin packages
+            if hub_pypi_requirements:
+                file_name = ".zenml_hub_pypi_requirements"
+                file_contents = "\n".join(hub_pypi_requirements)
+                requirements_files.append((file_name, file_contents))
+                if log:
+                    logger.info(
+                        "- Including hub requirements from PyPI: %s",
+                        ", ".join(f"`{r}`" for r in hub_pypi_requirements),
+                    )
+
         return requirements_files
 
+    @staticmethod
+    def _get_hub_requirements(
+        required_hub_plugins: List[str],
+    ) -> Tuple[Dict[str, List[str]], List[str]]:
+        """Get package requirements for ZenML Hub plugins.
+
+        Args:
+            required_hub_plugins: List of hub plugin names in the format
+                `(<author_username>/)<plugin_name>(==<version>)`.
+
+        Returns:
+            - A dict of the hub plugin packages themselves (which need to be
+                installed from a custom index, mapping index URLs to lists of
+                package names.
+            - A list of all unique dependencies of the required hub plugins
+                (which can be installed from PyPI).
+        """
+        from zenml._hub.client import HubClient
+        from zenml._hub.utils import parse_plugin_name, plugin_display_name
+
+        client = HubClient()
+
+        _DICT_TYPE = Union[Dict[str, List[str]], DefaultDict[str, List[str]]]
```

Comment:
Wow that is ugly :D

Why is this necessary, so you can return a `Dict` instead of `DefaultDict`? Maybe if we just do `return dict(hub_internal_requirements)` instead of re-assigning `hub_internal_requirements=dict(hub_internal_requirements)` we can avoid that?

Reply:
Yeah I agree, changed it as suggested 👍